test/unit/mixins/evented.test.js

```diff
@@ -504,3 +504,40 @@ QUnit.test('off() can remove a listener from an array of events on a different t
     target: b.eventBusEl_
   });
 });
+
+QUnit.test('on() with the same function on two different targets should be disposed correctly', function(assert) {
+  const a = this.targets.a = evented({});
+  const b = this.targets.b = evented({});
+  const c = this.targets.c = evented({});
+  let i = 0;
+  const update = () => {
+    i++;
+  };
+
+  a.on(b, 'foo', update);
+  a.on(c, 'bar', update);
+
+  // sanity test with b trigger
+  b.trigger('foo');
+  assert.strictEqual(i, 1, 'b trigger works');
+
+  // sanity test with c trigger
+  c.trigger('bar');
+  assert.strictEqual(i, 2, 'c trigger works');
+
+  b.trigger('dispose');
```

Comment:
This also causes `removeListenerOnDispose` for `c`'s to be removed for the `update` function. So when `a` disposes we won't run that function and the `foo` listener  will still exist.

Reply:
This happens because we set `removeListenerOnDispose`'s `guid` to the `guid` of the initial listener. so all `removeListenerOnDispose`s for the `update` function will have the same `guid` and they will all be removed.